packages/agent/src/agent/http/index.ts

```diff
@@ -407,63 +417,167 @@ export class HttpAgent implements Agent {
+      if (!this.#verifyQuerySignatures) {
+        resolve();
+      }
+      const subnetStatus = this.#subnetKeys.get(canisterId.toString());
+      if (subnetStatus) {
+        resolve(subnetStatus);
+      } else {
+        this.fetchSubnetKeys(canisterId)
+          .then(response => {
+            resolve(response);
+          })
+          .catch(error => {
+            reject(error);
+          });
+      }
     });
+    const [query, subnetStatus] = await Promise.all([queryPromise, subnetStatusPromise]);
+    // Skip verification if the user has disabled it
+    if (!this.#verifyQuerySignatures) {
+      return query;
+    }
+    return this.#verifyQueryResponse(query, subnetStatus);
+  }
 
-    // Apply transform for identity.
-    transformedRequest = await id?.transformRequest(transformedRequest);
+  /**
+   * See https://internetcomputer.org/docs/current/references/ic-interface-spec/#http-query for details on validation
+   * @param queryResponse - The response from the query
+   * @param subnetStatus - The subnet status, including all node keys
+   * @returns ApiQueryResponse
+   */
+  #verifyQueryResponse = (
+    queryResponse: ApiQueryResponse,
+    subnetStatus: SubnetStatus | void,
+  ): ApiQueryResponse => {
+    if (!subnetStatus || this.#verifyQuerySignatures === false) {
+      // This should not be called if the user has disabled verification
+      return queryResponse;
+    }
+    const { status, signatures, requestId } = queryResponse;
+
+    const domainSeparator = new TextEncoder().encode('\x0Bic-response');
+    signatures?.forEach(sig => {
+      const { timestamp } = sig;
+      let hash: ArrayBuffer;
+
+      // Hash is constructed differently depending on the status
+      if (status === 'replied') {
+        const { reply } = queryResponse;
+        hash = hashOfMap({
+          status: status,
+          reply: hashOfMap(reply),
+          timestamp: BigInt(timestamp),
+          request_id: requestId,
+        });
+      } else if (status === 'rejected') {
+        const { reject_code, reject_message, error_code } = queryResponse;
+        hash = hashOfMap({
+          status: status,
+          reject_code: reject_code,
+          reject_message: reject_message,
+          error_code: error_code,
+          timestamp: BigInt(timestamp),
+          request_id: requestId,
+        });
+      } else {
+        throw new Error(`Unknown status: ${status}`);
+      }
 
-    const body = cbor.encode(transformedRequest.body);
-    const response = await this._requestAndRetry(() =>
-      this._fetch('' + new URL(`/api/v2/canister/${canister.toText()}/query`, this._host), {
-        ...this._fetchOptions,
-        ...transformedRequest.request,
-        body,
-      }),
-    );
+      const separatorWithHash = concat(domainSeparator, new Uint8Array(hash));
 
-    const queryResponse: QueryResponse = cbor.decode(await response.arrayBuffer());
+      // FIX: check for match without verifying N times
```

Comment:
Are we going to fix this in this PR?

Reply:
Can you explain further? I don't recall that part of the spec

Reply:
Here we try all the node keys stored in `subnetStatus` to see if one of them can match the signature. 
The comment you left  `FIX: check for match without verifying N times` is from the previous PR about fetching node keys.
In the previous PR , we flatten the forks under `/subnet/<subnet_id>/node` and get a list of public keys. Because the path `/subnet/<subnet_id>/node/<node_id>/public_key` also contains the information of node id, we can store the mapping of `node-id->public key` in the `subnetStatus`. Then, instead of trying all keys, we just get it the from mapping by the `sig.identity`(the node_id which generated the response).

Reply:
This is fixed in https://github.com/dfinity/agent-js/pull/784/commits/fba09bf956943688625bcc02c066848e63ce3478